src/main/java/com/cryptomorin/xseries/particles/ParticleDisplay.java

```diff
@@ -1794,4 +1833,221 @@ public Vector mul(Vector point) {
             return new Vector(vx, vy, vz);
         }
     }
+
+    public interface ParticleData {
+        /**
+         * Returns true if the data type uses the offset fields,
+         * and false if it uses the data field.
+         * Sometimes depends on the particle and version.
+         */
+        public default boolean isOffsetBased(XParticle particle) {
+            return false;
+        }
+
+        public default double[] offsetValues() {
+            return new double[]{0, 0, 0};
+        }
+
+        public Object data(XParticle particle, double extra);
```

Comment:
I think it's better to just accept a ParticleDisplay here. Newer particles might require other data in the future.